### PR TITLE
ARROW-17051: [C++] Link Flight/gRPC/Protobuf consistently

### PR DESCRIPTION
**Pull request by lidavidm**
If Protobuf/gRPC are used statically, Flight must be as well, or else we can get odd runtime behavior due to the global state in those libraries when Flight SQL is involved (as Flight SQL would then bundle a second copy of Protobuf into its shared library).

### CODE REVIEW

**Review by raulcd — COMMENTED** [review]
Should we also change the travis jobs that fail?
```
-- If static Protobuf or gRPC are used, Arrow must be built statically
-- (These libraries have global state, and linkage must be consistent)
CMake Error at src/arrow/flight/CMakeLists.txt:43 (message):
  Must build Arrow statically to link Flight tests statically
```
https://github.com/apache/arrow/blob/master/.travis.yml#L96 and https://github.com/apache/arrow/blob/master/.travis.yml#L148

File: docker-compose.yml [hunk 1]
```diff
@@ -487,7 +487,9 @@ services:
       <<: *ccache
       CC: clang-${CLANG_TOOLS}
       CXX: clang++-${CLANG_TOOLS}
-      ARROW_BUILD_STATIC: "OFF"
+      # ARROW-17051: this build uses static Protobuf, so we must also
+      # use static Arrow to run Flight/Flight SQL tests
+      ARROW_BUILD_STATIC: "ON"
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply docker-compose.yml]: I'll move it. 22.04 is also too old (3.12, we want 3.15): https://packages.ubuntu.com/jammy/libprotobuf-dev

**Comment by lidavidm** [comment]
> Should we also change the travis jobs that fail?

Updated, thanks for catching that

**Comment by lidavidm** [comment]
Ok, I think this is ready. The CI failures seem to be incidental or addressed in other issues. Thanks Raúl and Kou for all the help here.

**Comment by kou** [comment]
It seems that `-ldl` is missing in Travis CI builds:

e.g.: https://app.travis-ci.com/github/apache/arrow/jobs/576693121#L2786

```text
FAILED: debug/plasma-store-server 
: && /usr/bin/c++  -Wno-noexcept-type -Wno-subobject-linkage  -fdiagnostics-color=always -ggdb -O0  -Wall -Wno-conversion -Wno-sign-conversion -Wunused-result -Werror -fno-semantic-interposition  -fPIC -g   src/plasma/CMakeFiles/plasma-store-server.dir/Unity/unity_0_cxx.cxx.o src/plasma/CMakeFiles/plasma-store-server.dir/Unity/unity_0_c.c.o src/plasma/CMakeFiles/plasma-store-server.dir/dlmalloc.cc.o  -o debug/plasma-store-server  debug/libplasma.a  debug/libarrow.a  /usr/lib/s390x-linux-gnu/libgflags.so.2.2.2  /usr/lib/s390x-linux-gnu/libssl.so  /usr/lib/s390x-linux-gnu/libcrypto.so  /usr/lib/s390x-linux-gnu/libbrotlienc.so  /usr/lib/s390x-linux-gnu/libbrotlidec.so  /usr/lib/s390x-linux-gnu/libbrotlicommon.so  /usr/lib/s390x-linux-gnu/libbz2.so  /usr/lib/s390x-linux-gnu/liblz4.so  /usr/lib/s390x-linux-gnu/libsnappy.so.1.1.8  /usr/lib/s390x-linux-gnu/libz.so  /usr/lib/s390x-linux-gnu/libzstd.so  opentelemetry_ep-install/lib/libopentelemetry_exporter_ostream_span.a  opentelemetry_ep-install/lib/libopentelemetry_exporter_otlp_http.a  opentelemetry_ep-install/lib/libopentelemetry_otlp_recordable.a  opentelemetry_ep-install/lib/libopentelemetry_trace.a  opentelemetry_ep-install/lib/libopentelemetry_resources.a  opentelemetry_ep-install/lib/libopentelemetry_common.a  opentelemetry_ep-install/lib/libopentelemetry_exporter_otlp_http_client.a  opentelemetry_ep-install/lib/libopentelemetry_proto.a  protobuf_ep-install/lib/libprotobuf.a  opentelemetry_ep-install/lib/libopentelemetry_http_client_curl.a  /usr/lib/s390x-linux-gnu/libcurl.so  /usr/lib/s390x-linux-gnu/libutf8proc.so  /usr/lib/s390x-linux-gnu/libre2.so  jemalloc_ep-prefix/src/jemalloc_ep/dist//lib/libjemalloc_pic.a  -pthread  -lrt  -lpthread && :
/usr/bin/ld: debug/libarrow.a(unity_6_cxx.cxx.o): undefined reference to symbol 'dlsym@@GLIBC_2.2'
/usr/bin/ld: /lib/s390x-linux-gnu/libdl.so.2: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```

Could you try this patch because `cpp/src/arrow/io/hdfs_internal.cc` uses `dlsym()`?

```diff
diff --git a/cpp/CMakeLists.txt b/cpp/CMakeLists.txt
index b67f90e0bd..945ff7b6f8 100644
--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -862,7 +862,7 @@ add_dependencies(arrow_test_dependencies toolchain-tests)
 
 if(ARROW_STATIC_LINK_LIBS)
   add_dependencies(arrow_dependencies ${ARROW_STATIC_LINK_LIBS})
-  if(ARROW_ORC)
+  if(ARROW_HDFS OR ARROW_ORC)
     if(NOT MSVC_TOOLCHAIN)
       list(APPEND ARROW_STATIC_LINK_LIBS ${CMAKE_DL_LIBS})
       list(APPEND ARROW_STATIC_INSTALL_INTERFACE_LIBS ${CMAKE_DL_LIBS})
```

**Comment by lidavidm** [comment]
Done - thanks for catching that!

**Comment by pitrou** [comment]
Remotely related: https://github.com/substrait-io/substrait/issues/249

**Review by raulcd — APPROVED** [review]
Thanks @lidavidm

File: .travis.yml [hunk 2]
```diff
@@ -93,7 +93,7 @@ jobs:
         # aws-sdk-cpp.
         DOCKER_RUN_ARGS: >-
           "
-          -e ARROW_BUILD_STATIC=OFF
+          -e ARROW_BUILD_STATIC=ON
```
Comment (kou) [inline .travis.yml:96]: Can we remove this because we have `ARROW_BUILD_STATIC=OFF` in Dockerfile?

File: .travis.yml [hunk 3]
```diff
@@ -145,7 +145,7 @@ jobs:
...
```
Comment (kou) [inline .travis.yml:148]: Can we remove this because we have `ARROW_BUILD_STATIC=OFF` in Dockerfile?

File: docker-compose.yml [hunk 4]
```diff
@@ -1214,6 +1213,7 @@ services:
     shm_size: *shm-size
     environment:
       <<: *ccache
+      ARROW_BUILD_TESTS: 'OFF'
```
Comment (kou) [inline docker-compose.yml:1216]: Is this needed?
`ci/docker/linux-apt-r.dockerfile` has `ARROW_BUILD_TESTS=OFF`.

**Comment by lidavidm** [comment]
Seems CI is still pending, but any other feedback here? (Thanks Kou for the comments so far.)

**Review by kou — APPROVED** [review]
+1

**Pull request merged**